loadbalancing.md

```diff
@@ -0,0 +1,345 @@
+
+It is also possible to use DNS round-robin. However this is not advisable as this won't prevent connections from going to a machine that is unavailable. This means that your application may have to re-try connections more often if not all servers are available.
+
+Some of the newer MySQL connectors have support for [DNS SRV](https://dev.mysql.com/doc/refman/8.0/en/connecting-using-dns-srv.html). This is somewhat similar to round-robin DNS, but allows you to set priorities. The benefit from this is that it is an industry standard and that this makes the client-side configuration easier.
+
+If you are using Kubernetes you probably want to use the [LoadBalancer](https://docs.pingcap.com/tidb-in-kubernetes/stable/access-tidb#loadbalancer) service type that works with Amazon ELB and similar services from other cloud vendors. For on-premise Kubernetes something like [metallb](https://metallb.universe.tf/) can be used.
+
+When your are not using Kubernetes to deploy on a cloud service you can still use the loadbalancing service from your cloud provider.
+
+Another common option is to use a software based loadbalancer. Common ones are ProxySQL, haproxy and MySQL Router.
+
+The last type of loadbalancing that we will discuss here is hardware based loadbalancers. These are physical machines where you plug in a network cable. These are often costly, but also offer a high throughput.
+
+## General loadbalancing requirements for TiDB
+
+Where loadbalancers may offer advanced features like read/write splitting for MySQL, this is not needed for TiDB. The loadbalancer does not need to be able to inspect the MySQL protocol. A TCP level loadbalacer works fine for TiDB.
+
+The loadbalancer only needs access to the TiDB servers (TCP port 4000 by default) and doesn't need access to PD or TiKV servers. When using the HTTP status access to TCP port 10080 is also required.
+
+It is recommended to use `http://<ip_of_tidb>:10080/status` as the health check since it supports the use of the _graceful shutdown_ feature. This feature allows you to first drain client connections before shutting down a TiDB Server and reduces the impact on applications. The [`graceful-wait-before-shutdown`](/tidb-configuration-file.md#graceful-wait-before-shutdown-new-in-v50) variable needs to be set to a non-zero value to benefit from this feature.
+
+Health checks that check for `read_only` or other MySQL or InnoDB variables should be avoided.
```

Comment:
- Is it common to do the health check by accessing system variables?(I know little about this)

- If using system variables like `read_only` is not recommended, it's better to add an explanation here, for example, "because they are [no-op](https://docs.pingcap.com/tidb/stable/system-variables#tidb_enable_noop_functions-new-in-v40) in TiDB'?"

Reply:
ProxySQL uses `read_only` to move servers between read and write pools: https://proxysql.com/documentation/backend-monitoring/

[this](https://severalnines.com/resources/database-management-tutorials/mysql-load-balancing-haproxy-tutorial) tutorial from severalnines also has a script that uses `read_only` to check if a master is "healty" or not.

For Galera based clusters [`wsrep_ready`](https://www.percona.com/doc/percona-xtradb-cluster/5.7/wsrep-status-index.html#wsrep_ready) is often used.

Besides this using [`Innodb_buffer_pool_load_status`](https://dev.mysql.com/doc/refman/8.0/en/server-status-variables.html#statvar_Innodb_buffer_pool_load_status) is also a common practice to only allow servers to enter a pool once they finished loading the pages that were in the buffer pool before the shutdown. This is to avoid cold caches.